daffodil-lib/src/main/scala/org/apache/daffodil/lib/api/Diagnostic.scala

```diff
@@ -308,4 +308,12 @@ trait WithDiagnostics {
    * then one can proceed to run the compiled entity.
    */
   def isError: Boolean
+
+  /**
+   * Helper method to check that isError is false, if not it throws
+   * a usage error caused by illegal state caused by a compilation error
+   */
+  def checkNotError(): Unit = {
+    Assert.usage(!isError, new IllegalStateException(getDiagnostics.find(_.isError).get))
```

Comment:
We've lost the message that says the API is being used incorrectly so a developer using the API wont know what to change. The user will just see something like:

> Usage Error: InvalidStateException, cause: Schema Defition Error: ....

The SDE found in the diagnostic is not the problem, it's more like a symptom. The underlying problem is incorrect API usage, so that needs to be included in the error message. The core issue of this bug is that it's not clear what to do if the check fails and the user gets an exception, and I don't think this does anything to make that more clear.

Reply:
We can pass the constructor of IllegalStateException both a message string AND a cause.
Then the string can explain that you must call 'isError' to ensure there are not errors, but the error creating that situation is still featured also.

Reply:
How's changing the definition to 

```scala
  def checkNotError(): Unit = {
    Assert.usage(!isError, new IllegalStateException(
      "Must call isError() to ensure there are no errors",
      getDiagnostics.find(_.isError).get,
    ))
  }
```

Reply:
The new cause will be:
```
cause = {IllegalStateException@2841} "java.lang.IllegalStateException: Must call isError() to ensure there are no errors"
 detailMessage = "Must call isError() to ensure there are no errors"
 cause = {SchemaDefinitionError@2853} Schema Definition Error: Error loading schema due to org.xml.sax.SAXParseException; systemId: file:/home/okilo@columbia.tresys.com/git/owlsvcs/data/tools/daffodil/daffodil-sapi/target/scala-2.12/test-classes/test/sapi/mySchema6.dfdl.xsd; lineNumber: 32; columnNumber: 74; src-resolve: Cannot resolve the name 'tns:nonExistent' to a(n) 'type definition' component.\nSchema context: file:/home/okilo@columbia.tresys.com/git/owlsvcs/data/tools/daffodil/daffodil-sapi/target/scala-2.12/test-classes/test/sapi/mySchema6.dfdl.xsd Location in .../test-classes/test/sapi/mySchema6.dfdl.xsd
 ```

Reply:
Seems reasonable to me.

I forget why we added `IllegalStateException`, but maybe we don't even need it anymore? Now that `UsageException` accepts a cause maybe this is just:

```
 def checkNotError(): Unit = {
    Assert.usage(!isError,
      "Must call isError() to ensure there are no errors",
      getDiagnostics.find(_.isError).get,
    )
  }
```

I don't feel strongly about it. Either works for me if having `IllegalStateException` provides extra useful information.

Reply:
I think illegal state adds the information that the usage is not that something is wrong with an argument, it's that you called it in the wrong state. Differentiates from IllegalArgumentException.